framework/aster-frame/src/vm/page_table/properties.rs

```diff
@@ -0,0 +1,205 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use core::fmt::Debug;
+
+use pod::Pod;
+
+use crate::vm::{Paddr, Vaddr, VmPerm};
+
+/// A minimal set of constants that determines the flags of the page table.
+/// This provides an abstraction over most paging modes in common architectures.
+pub trait PageTableConstsTrait: Debug {
+    /// The smallest page size.
+    const BASE_PAGE_SIZE: usize;
+
+    /// The number of levels in the page table.
+    /// The level 1 is the leaf level, and the level `NR_LEVELS` is the root level.
+    const NR_LEVELS: usize;
+
+    /// The highest level that a PTE can be directly used to translate a VA.
+    /// This affects the the largest page size supported by the page table.
+    const HIGHEST_TRANSLATION_LEVEL: usize;
+
+    /// The size of a PTE.
+    const ENTRY_SIZE: usize;
+}
+
+// Here are some const values that are determined by the page table constants.
+
+/// The number of PTEs per page table frame.
+pub const fn nr_entries_per_frame<C: PageTableConstsTrait>() -> usize {
+    C::BASE_PAGE_SIZE / C::ENTRY_SIZE
+}
+
+/// The number of bits used to index a PTE in a page table frame.
+pub const fn in_frame_index_bits<C: PageTableConstsTrait>() -> usize {
+    nr_entries_per_frame::<C>().ilog2() as usize
+}
+
+/// The index of a VA's PTE in a page table frame at the given level.
+pub const fn in_frame_index<C: PageTableConstsTrait>(va: Vaddr, level: usize) -> usize {
+    va >> (C::BASE_PAGE_SIZE.ilog2() as usize + in_frame_index_bits::<C>() * (level - 1))
+        & (nr_entries_per_frame::<C>() - 1)
+}
+
+/// The page size at a given level.
+pub const fn page_size<C: PageTableConstsTrait>(level: usize) -> usize {
+    C::BASE_PAGE_SIZE << (in_frame_index_bits::<C>() * (level - 1))
+}
+
+bitflags::bitflags! {
+    /// The status of a memory mapping recorded by the hardware.
+    pub struct MapStatus: u32 {
+        const ACCESSED = 0b0000_0001;
+        const DIRTY    = 0b0000_0010;
```

Comment:
This is architectural dependent. So it should be moved to `arch/`.

Reply:
No it is not architectural dependent :>